lib/src/flutter_downloader.dart

```diff
@@ -0,0 +1,117 @@
+
+part 'legacy_api.dart';
+
+/// The target where the download should be visible for the user.
+enum DownloadTarget {
+  /// Put the download in the download folder. If the target platform does
+  /// support download folders internal will be used.
+  downloadsFolder,
+
+  /// Put the download in the desktop folder. If the target platform does
+  /// support download folders internal will be used.
+  desktopFolder,
+
+  /// Put the download into the internal storage of the app.
+  internal,
+}
+
+/// Abstraction to observe download status updates
+abstract class DownloadProgress extends ChangeNotifier {
+  /// The state of the download
+  DownloadStatus get status;
+
+  /// The current progress in per mille [0...1000]
+  // If you need percent simply device it by ten. Per mille is used here for
+  // big downloads where you would see big jumps between progress steps.
+  int get progress;
+}
+
+/// Factory interface
+typedef CustomHttpClientFactory = HttpClient Function();
+
+/// Provides access to all functions of the plugin in a single place.
+/// The change notifier should just inform about changes of the
+/// [DownloadProgress].
+class FlutterDownloader extends ChangeNotifier
+    with FlutterDownloaderLegacy
+    implements DownloadProgress {
+  /// Get the flutter downloader.
+  factory FlutterDownloader() {
+    // Singleton is in Legacy part since it is required there too
+    return FlutterDownloaderLegacy._singleton;
+  }
+
+  FlutterDownloader._internal();
+
+  /// Add a custom http factory for platforms without platform specific
+  /// implementations like Android and iOS.
+  static CustomHttpClientFactory? customHttpClientFactory;
+
+  @override
+  int get progress => throw UnimplementedError();
+
+  @override
+  DownloadStatus get status => throw UnimplementedError();
+
+  // TODO check if that is required in the future
+  static bool _debug = false;
+
+  /// If true, more logs are printed.
+  static bool get debug => _debug;
+
+  /// Start a new download. The [Download] instance encapsulates the download
+  /// status and controls like cancel, pause and resume.
+  Future<Download> startDownload(
+    String url, {
+    String userAgent = 'flutter_downloader',
+    String? fileName,
+    Map<String, String> additionalHeaders = const {},
+    DownloadTarget target = DownloadTarget.internal,
+  }) async {
+    final headers = Map<String, String>.from(additionalHeaders);
+    headers['User-Agent'] = userAgent;
+    final download = await PlatformDownload.create(
+      url: url,
+      headers: headers,
+      target: target,
+    );
+    unawaited(download.resume());
+    return download;
+  }
+
+  /// Returns all known downloads.
+  Future<List<Download>> getDownloads() async {
+    final baseDir = await PlatformDownload.getLocalDir();
+    final files = await baseDir.list().toList();
+    final downloads = <Download>[];
+    for (final file in files) {
+      //print('checking File: ${file.path}');
```

Comment:
Let's make `FlutterDownloader` accept an optional function argument called `logger`. By default it's null, but the user can pass whatever they want to get logs from the internals of FlutterDownloader.